Refactor to use vite. Manage JS dependencies with NPM.

UFO-101 opened the pull request.

UFO-101 commented:
Thanks for the comments. I don't have time to update today, but should be able to merge tomorrow.

> What's the reason you broke out the JVM and JS code into separate "shared" and non-shared modules

Not much reason, happy to change that back. I was trying to stick to Mill's default module layouts.

> Similarly, why have the js/jvm package distinction in the code? 

When I separated the modules I needed some way to have the code in the `package object` available everywhere.

julianmichael commented:
Makes sense! Yeah I think it's ultimately simpler with just two modules the way I did it originally. Defining the package object is a bit odd (you have a package object in the shared code which extends a trait — `PackagePlatformExtensions` — defined in platform-specific code) but actually works out nicely. Happy to run through it and do a quick cleanup once you're done.

UFO-101 commented:
> Yeah I think it's ultimately simpler with just two modules the way I did it originally. Defining the package object is a bit odd (you have a package object in the shared code which extends a trait — PackagePlatformExtensions — defined in platform-specific code) but actually works out nicely. 

- I've reverted the module structure so that there are just two modules, both including `shared/` in their source.
- I've reverted the package structure so now both modules just use the `debate` package.

There are a couple of issues to be worked out:
 - Vite complains that it can't find all the source maps (I think this is not a new problem, as I noticed it before).
 - Vite's hot module reloading is not working (the page just refreshes when the source updates).

However I think it makes sense to merge it asap anyway because it's still easier to work with than before. And every other change will conflict with this one, so it would save some git wrangling. I will resolve the above issues in a future PR.

julianmichael commented:
Awesome, I will try out the branch and work on fixing conflicts as soon as I can make time.

julianmichael reviewed (changes requested):
Looks good! Pushed some changes I needed to make to get it to work on my end. One last thing to do is set it up to work behind a single server endpoint, which will let me easily deploy it in the prod environment (eg i can put it behind a single reverse proxy). I'm looking at trying to get it to work using Vite settings (eg [here](https://vitejs.dev/config/server-options.html#server-proxy)). Haven't managed it after a few minutes of trying, feel free to give it a go. My current attempt is in [this branch](https://github.com/julianmichael/debate/tree/vite-refactor-proxy), maybe give it a go?

julianmichael commented:
> @julianmichael would you like me to review this as well?

I think it looks good overall, see my previous comment for the last change I think we need for deployment, lmk if you want to give it a go

smithjessk commented:
> > @julianmichael would you like me to review this as well?
> 
> I think it looks good overall, see my previous comment for the last change I think we need for deployment, lmk if you want to give it a go

@julianmichael I'll pass on figuring out the settings. But I'll do a quick spot-check review!

smithjessk commented on README.md:
```diff
@@ -3,27 +3,37 @@ Web interface for information-asymmetric debates.
-This will host your server by HTTP on port 8080 and save all of the relevant server state under
-`save/`.
+For development, you must run 3 commands in separate terminals in the root directory of the project:
+ - `mill --no-server -j 8 -w js.publcDev` to start incremental compilation of the frontend.
```
@UFO-101 why `-j 8` instead of `-j 0`?

UFO-101 replied:
Oops, thanks for pointing that out! I just didn't want it to use up my whole cpu.

smithjessk commented on js/src/ConnectedLobbyPage.scala:
```diff
@@ -36,12 +36,12 @@ object ConnectedLobbyPage {
   ): String = {
     val prefix = if (isOfficial) "official" else "practice"
-    s"${Helpers.wsProtocol()}//${dom.document.location.host}/$prefix-ws/$roomName?name=$participantId"
+    s"${Helpers.wsProtocol()}//${dom.document.location.hostname}:8080/$prefix-ws/$roomName?name=$participantId"
```
Maybe it's better to not hard-code `8080`. It could go in `Helpers`.

UFO-101 replied:
Agreed, will change.

smithjessk commented on js/src/ConnectedLobbyPage.scala:
```diff
@@ -36,12 +36,12 @@ object ConnectedLobbyPage {
   val httpProtocol = dom.document.location.protocol
   val qualityApiUrl: String = {
-    s"$httpProtocol//${dom.document.location.host}/$qualityServiceApiEndpoint"
+    s"$httpProtocol//${dom.document.location.hostname}:8080/$qualityServiceApiEndpoint"
```
Same comment as above about hard-coding `8080`.

smithjessk commented on js/src/ConnectedLobbyPage.scala:
```diff
@@ -140,7 +140,7 @@ object ConnectedLobbyPage {
               S.simpleSelected.when(isCurrent)
             )(
-              <.div(S.optionTitle)("Facilitators"),
+              <.div(S.optionTitle)("88838 Facilitators"),
```
What's `88838 Facilitators`?

UFO-101 replied:
Ooops, left in from debuging 😄

smithjessk reviewed (commented):
Just left some general questions. Didn't do an exhaustive review. Thanks for the work!

julianmichael commented:
Followup on my earlier request — in retrospect it's probably easier to have the Scala server serve the static files in prod (like it is in the current build), that way we only need to start up the one server. The dev and prod configurations would be fairly different but I think it's still probably the easiest way.

julianmichael commented:
I'm going to go ahead and merge this now so I don't keep creating conflicts as I update main.
